### DISCUSSION TITLE
KoScope / KoFile from source code / git

### DISCUSSION THREAD

**Discussion by minogin** [LemonAppDev/konsist#586] [Ask a Question]
### Wanted end result.

We use Konsist in a way that we are interested in which git commits introduced failing konsist tests. Now we are at a point where it would be very convenient to get a konsist file / scope from git or at least from a source code string. What's the best way to do that?

### What you tried so far.

-

**Comment by igorwojda** [comment]
Happy to hear that you use it already :-)
Please explain your scenario in more detail.

**Reply by minogin** [reply]
We implemented some konsist tests and linked them to jgit so that they only fail on recent changes (commits). Otherwise we need to fix all the legacy code at once which is too much work.
Now we see that even this is too strict and we want to make it that konsist tests fail only on recently added/modified methods. Probably that's not a konsist task at all. But I am now thinking on possible implementation. One of the options is too get parent commit and check if the tests was failing already. Then we can say that the change didn't make it worse :)

**Reply by minogin** [reply]
Also want to thank you for this great library as this is something we were actually missing for the long time.

**Reply by minogin** [reply]
So, anyways one option which Id like to see is creating a KoScope out of raw source fetched from git. Not from file system.

**Reply by igorwojda** [reply]
Interesting case.

Heaving list of modified files you could create a scope individual for each file
https://docs.konsist.lemonappdev.com/writing-tests/koscope#file-scope

Then you could use scope composition to add these individual scopes and create a scope containing all changed files
https://docs.konsist.lemonappdev.com/writing-tests/koscope#scope-composition

Let me know if this works for you.

BTW Structural checks should work fine, however, architecture checks may now work with this as an empty layer will throw an exception.

BTW2 If this works we could consider adding `fun scopeFromFile(paths: List<String>): KoScope` method to the API

**Reply by minogin** [reply]
I'd say something like `fun scopeFromSources(fileSourceCodes: List<String>): KoScope` would be nice

**Reply by minogin** [reply]
You may also consider integration with jgit as running Konsist over our project atm we get around 1500 failing classes which makes devs a bit... nervous :)
I can share some insights with you as well.

**Reply by igorwojda** [reply]
I have no idea what the `jgit` is, but I am open to learning more. Please explain this in great detail.

**Reply by igorwojda** [reply]
The `scopeFromFiles` will be available in the next release (https://github.com/LemonAppDev/konsist/pull/588 and follow up https://github.com/LemonAppDev/konsist/pull/589).
Let's see how this one goes.

**Comment by minogin** [comment]
So, regarding jgit - it's just java API for the git.
What we do exactly:
1. Test the classes with Konsist 
2. For those which have issues find the last git commit which introduced them using jgit
3. If this commit date is after e.g. 01.09 we throw an error
So the idea is that you don't need to fix all the issues in the project, but if you touch some legacy file you will have to fix it.
And of course if you introduce new issues as well.

**Reply by minogin** [reply]
Sample code:
```
@Test
    fun `controller does not depend on repository`() {
        controllers
            .flatMapWith { it.imports() }
            .filter { (_, i) -> i.fromRepositoryPackage() }
            .filter { (c, _) -> c.recentlyModified() }     // Checks whether the last git commit where controller was modified is after 01 Sep 2023
            .firstOrNull()
            ?.let { (c, i) ->
                throw AssertionError(
                    "${c.fullyQualifiedName} depends on ${i.name}\n" +
                            "Controller should not depend on repository layer\n${blame(c)}" //  and blames the author of that commit
                )
            }
    }
    ```

**Comment by igorwojda** [comment]
A new version v0.13.0 has been released. It contains the fix.
See https://github.com/LemonAppDev/konsist/releases/tag/v0.13.0

I will close this discussion and merge it with another one where I have provided a high-level detailed answear.
Let's continue here https://github.com/LemonAppDev/konsist/discussions/649#discussioncomment-7229644
